### ISSUE TITLE
[@typescript-eslint/no-unnecessary-type-assertion] No Unnecessary Assertion False Positive

### ISSUE THREAD

**Issue by hbiede** [typescript-eslint/typescript-eslint#3558]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

[Minimal Repo to Reproduce](https://github.com/hbiede/test-eslint)

.eslintrc:
```JSON
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "**/tsconfig.json",
    "ecmaVersion": 2018,
    "ecmaFeatures": {
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": ["@typescript-eslint"],
  "extends": [
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking"
  ],
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    }
  }
}
```

index.ts
```TS
const fontSizes = {
  h1: 50,
  h2: 45,
  h3: 40,
  h4: 35,
  h5: 30,
  h6: 25,
}

type Props = {
  level: 1 | 2 | 3 | 4 | 5 | 6;
}
type HeadingTag = `h${Props['level']}`;

const getFontSizeForHeader = ({ level }: Props) => fontSizes[`h${level}` as HeadingTag];

console.log('Test', getFontSizeForHeader({ level: 1 }));
```

tsconfig.json
```JSON
{
  "compilerOptions": {
    "strict": true
  },
  "include": ["src/*"],
}
```

**Expected Result**

No errors. I need to assert the template string to be `HeadingTag` because otherwise I get a TS error for attempting to index the object with a `string`

**Actual Result**

```
./src/index.ts
  15:62  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
```

**Additional Info**

[Gist of ESLint debug log](https://gist.github.com/hbiede/be88db334315661d121d759f851b4362)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.27.0` |
| `@typescript-eslint/parser`        | `4.27.0` |
| `TypeScript`                       | `4.3.4` |
| `ESLint`                           | `7.29.0` |
| `node`                             | `12.18.4` |

**Comment by bradzacher** [comment]
The problem here is that TS types `` `h${level}` `` as `'h1' | 'h2' | ...`, but when it then uses it in the index it widens the type to `string` for some reason.
I don't think there's anything we can do about this. When we inspect the types the are equal, but when TS does the check it widens it.

You can use `as const` instead.
